### PR TITLE
[CC] [autodiff] Support AdStack on C backend

### LINKED ISSUE #1748: [AutoDiff] [CC] Add ti.extension.adstack support to C backend
**Concisely describe the proposed feature**
I'd like to add the support of adstack to the C backend.
Some of the advanced usage of autodiff seems depends on this feature very much, e.g., `mass_spring.py` in [difftaichi](https://github.com/yuanming-hu/difftaichi/tree/master/examples), as @samuela reported in https://github.com/taichi-dev/taichi/issues/1644#issuecomment-671617104.

**Describe the solution you'd like (if any)**
I'm not super clear about the autodiff system, so I feel #1257 hard for me to understand. @k-ye Would you mind provide me with more information?
- `StackPopStmt`
- `StackPushStmt`
- `StackLoadTopStmt`
- `StackLoadTopAdjStmt`
- `StackAccAdjointStmt`

What does `adjoint` and `primal` mean and their difference? It seems we have two 32-bit float values per-stack push? Is `Adj` a shortcut for `Adjoint`, or simply `adjust stack`?

**Additional comments**
@samuela's idea is to call Taichi kernels directly from Julia, that's why we need C backend instead of the LLVM backend, which doesn't have an export option for now.

### PR DESCRIPTION
**Pull request by archibate**
Related issue = close #1748

[[Click here for the format server]](http://kun.csail.mit.edu:31415/)

----

### CODE REVIEW

**Comment by archibate** [comment]
Can we confirm that this is an issue in autodiff system or C backend?

**Comment by xumingkuan** [comment]
> Can we confirm that this is an issue in autodiff system or C backend?

I'll take a look tomorrow.

**Comment by xumingkuan** [comment]
It's an issue in autodiff.

test case:
`test_ad_fibonacci_index()` in test_ad_for.py
log:
```
[I 08/26/20 19:11:30.900] [compile_to_offloads.cpp:taichi::lang::irpass::`anon
ymous-namespace'::make_pass_printer::<lambda_fe1d620add3df83d4ee306f9c0ab10ca>
::operator ()@18] [fib_c5_0_grad_grad] Simplified I:
kernel {
  <i32 x1> $0 = const [5]
  <i32 x1> $1 = const [0]
  <i32 x1> $2 = const [1]
  <i32 x1> $3 = const [10]
  $4 : for in range($1, $3) (vectorize 1) block_dim=adaptive {
    <i32 x1> $5 = loop $4 index 0
    <f32*x1> $6 = global ptr [S6place_f32], index [$5] activate=true
    <f32 x1> $7 = global load $6
    <f32*x1> $8 = global ptr [S10place_f32], index [] activate=true
    <f32 x1> $9 = atomic add($8, $7)
  }
  $10 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
    <i32 x1> $11 = alloca
    <i32 x1> $12 = alloca
    <i32 x1> $13 : local store [$12 <- $2]
    $14 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
      <i32 x1> $15 = local load [ [$12[0]]]
      <i32 x1> $16 = local load [ [$11[0]]]
      <i32 x1> $17 = add $16 $15
      <i32 x1> $18 : local store [$11 <- $15]
      <i32 x1> $19 : local store [$12 <- $17]
      <f32*x1> $20 = global ptr [S2place_f32], index [$17] activate=true
      <f32 x1> $21 = global load $20
      <f32*x1> $22 = global ptr [S6place_f32], index [$17] activate=true
      <f32 x1> $23 = atomic add($22, $21)
    }
  }
}
[I 08/26/20 19:11:30.903] [compile_to_offloads.cpp:taichi::lang::irpass::`anon
ymous-namespace'::make_pass_printer::<lambda_fe1d620add3df83d4ee306f9c0ab10ca>
::operator ()@18] [fib_c5_0_grad_grad] Gradient:
kernel {
  <i32 x1> $0 = const [5]
  <i32 x1> $1 = const [0]
  <i32 x1> $2 = const [1]
  <i32 x1> $3 = const [10]
  $4 : for in range($1, $3) (vectorize 1) block_dim=adaptive {
    <i32 x1> $5 = loop $4 index 0
    <f32*x1> $6 = global ptr [S6place_f32], index [$5] activate=true
    <f32*x1> $7 = global ptr [S10place_f32], index [] activate=true
    <f32*x1> $8 = global ptr [S12place_f32], index [] activate=true
    <f32 x1> $9 = global load $8
    <f32*x1> $10 = global ptr [S8place_f32], index [$5] activate=true
    <f32 x1> $11 = atomic add($10, $9)
  }
  $12 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
    <f32 x1> $13 = stack alloc (max_size=16)
    <i32 x1> $14 = stack alloc (max_size=16)
    <i32 x1> $15 = stack alloc (max_size=16)
    <i32 x1> $16 = stack alloc (max_size=16)
    <i32 x1> $17 : stack push $16, val = $2
    $18 : for in range($1, $0) (vectorize 1) block_dim=adaptive {
      <i32 x1> $19 = stack load top $16
      <i32 x1> $20 = stack load top $15 // <------------------------------ empty stack!
      <i32 x1> $21 = add $20 $19
      <i32 x1> $22 : stack push $14, val = $21
      <i32 x1> $23 = stack load top $14
      <i32 x1> $24 : stack push $15, val = $19
      <i32 x1> $25 : stack push $16, val = $23
      <f32*x1> $26 = global ptr [S2place_f32], index [$23] activate=true
      <f32 x1> $27 = global load $26
      <f32 x1> $28 : stack push $13, val = $27
      <f32*x1> $29 = global ptr [S6place_f32], index [$23] activate=true
    }
    $30 : reversed for in range($1, $0) (vectorize 1) block_dim=adaptive {
      <i32 x1> $31 = stack load top $14
      <f32*x1> $32 = global ptr [S8place_f32], index [$31] activate=true
      <f32 x1> $33 = global load $32
      <f32 x1> $34 : stack acc adj $13, val = $33
      <f32 x1> $35 = stack load top adj $13
      <f32 x1> $36 : stack pop $13
      <f32*x1> $37 = global ptr [S4place_f32], index [$31] activate=true
      <f32 x1> $38 = atomic add($37, $35)
      <i32 x1> $39 : stack pop $16
      <i32 x1> $40 : stack pop $15
      <i32 x1> $41 : stack pop $14
    }
    <i32 x1> $42 : stack pop $16
  }
}

```

**Comment by archibate** [comment]
Great, so will we merge this PR before or after that issue is resolved?

**Comment by xumingkuan** [comment]
`*n = 1;` looks too hacky to me... I prefer to merge this after that issue is resolved if that won't take too much time.

File: taichi/backends/cc/runtime/base.h
Comment (xumingkuan) [inline taichi/backends/cc/runtime/base.h:102]: After merging #1791: 
```suggestion
  *n = 0;
```

Note that zero-initialization is already handled in `Ti_ad_stack_push`.

**Review by xumingkuan — APPROVED** [review]
Cool! LGTM.

**Pull request merged**
